### PR TITLE
Bcc32 compatibility

### PR DESCRIPTION
**Pull request by starkmapper**
I did some work to get fmt working on the "classic" Embarcadero compiler.

### CODE REVIEW

**Comment by vitaut** [comment]
Hi @starkmapper, thanks for the PR!

Since this is your first contribution to {fmt}, please review [CONTRIBUTING.md](https://github.com/fmtlib/fmt/blob/ebec00138bd10e949ff34955bcb096387cd32acc/CONTRIBUTING.md), particularly the part about licensing, and comment here whether you agree with it being applied to your contributions including this PR.

**Comment by starkmapper** [comment]
I have read contributing.md and agree that my contribution will be licensed under the fmt license.

File: fmt/format.cc [hunk 1]
```diff
@@ -83,10 +83,12 @@ static inline fmt::internal::Null<> strerror_s(char *, std::size_t, ...) {
...
 namespace fmt {
 
-FMT_FUNC internal::RuntimeError::~RuntimeError() FMT_DTOR_NOEXCEPT {}
+namespace internal {
+FMT_FUNC RuntimeError::~RuntimeError() FMT_DTOR_NOEXCEPT {}
+} // namespace internal
+
 FMT_FUNC FormatError::~FormatError() FMT_DTOR_NOEXCEPT {}
 FMT_FUNC SystemError::~SystemError() FMT_DTOR_NOEXCEPT {}
-
```
Comment (vitaut) [inline fmt/format.cc]: Please avoid unnecessary whitespace changes.

File: fmt/format.h [hunk 2]
```diff
@@ -1331,11 +1331,11 @@ template <typename T, T> struct LConvCheck {
 // Returns the thousands separator for the current locale.
 // We check if ``lconv`` contains ``thousands_sep`` because on Android
 // ``lconv`` is stubbed as an empty struct.
-template <typename LConv>
+/*template <typename LConv>
 inline StringRef thousands_sep(
     LConv *lc, LConvCheck<char *LConv::*, &LConv::thousands_sep> = 0) {
   return lc->thousands_sep;
-}
+}*/
```
Comment (vitaut) [inline fmt/format.h]: This should be conditionally compiled instead of commented out.

File: fmt/format.h [hunk 3]
```diff
@@ -2138,8 +2138,10 @@ class ArgFormatterBase : public ArgVisitor<Impl, void> {
 
   // workaround MSVC two-phase lookup issue
   typedef internal::Arg Arg;
+  typedef Char          Char_Type;
+  typedef BasicWriter<Char> Writer_Type;
```
Comment (vitaut) [inline fmt/format.h]: `Char_Type` -> `CharType`, `Writer_Type` -> `WriterType` for consistency with other names.

File: fmt/format.h [hunk 4]
```diff
@@ -2432,6 +2433,19 @@ struct ArgArray<N, false/*IsPacked*/> {
   template <typename Formatter, typename T>
   static Arg make(const T &value) { return MakeArg<Formatter>(value); }
 };
+#else
+template <std::size_t N>
+struct ArgArray
+{
```
Comment (vitaut) [inline fmt/format.h]: nit: please follow the Google Style: https://google.github.io/styleguide/cppguide.html, in particular 

```c++
struct ArgArray {
```

File: fmt/format.cc (hunk 1 above)
Comment (starkmapper) [reply fmt/format.cc]: I did a sweep through all changes to resolve all whitespace issues. Should be OK now, please let me know if I missed anything.

File: fmt/format.h (hunk 2 above)
Comment (starkmapper) [reply fmt/format.h]: That was what commit eec3fcdeabc52f85fcb58a567fae0a6d5d700123 did :-)

File: fmt/format.h (hunk 3 above)
Comment (starkmapper) [reply fmt/format.h]: Should be fixed in 8915b2754397b5c4090267599d03cd31f1e323f0

File: fmt/format.h (hunk 4 above)
Comment (starkmapper) [reply fmt/format.h]: I did my best to adhere to the style guide. I'm not sure if single line functions are allowed, but if so, this should be ok now :-).

**Comment by starkmapper** [comment]
I didn't want to rebase+squash just yet, but let me know if that would help you.

**Comment by starkmapper** [comment]
I'm also a little worried about the CI failing, is there anything I can do about that?

**Comment by vitaut** [comment]
> I didn't want to rebase+squash just yet, but let me know if that would help you.

That would help.

> I'm also a little worried about the CI failing, is there anything I can do about that?

Please check the error at https://travis-ci.org/fmtlib/fmt/jobs/507890085. Looks like they are mostly namespace-related.

**Comment by starkmapper** [comment]
Squashed and fixed non-windows builds (waiting for travis to confirm :-) )

**Comment by starkmapper** [comment]
> I'm not sure what's going on with the appveyor build. Did I break that or is something else going on?
Tried the appveyor script locally and that seems to work.
image

File: fmt/format.h [hunk 5]
```diff
@@ -2469,74 +2498,6 @@ inline uint64_t make_type(FMT_GEN15(FMT_ARG_TYPE_DEFAULT)) {
...
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 3) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 4) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 5) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 6) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 7) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 8) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 9) \
-  FMT_CTOR(ctor, func, arg0_type, arg1_type, 10)
-#endif
-
```
Comment (vitaut) [inline fmt/format.h]: Why move this block?

File: fmt/format.h [hunk 6]
```diff
@@ -2593,11 +2687,8 @@ class SystemError : public internal::RuntimeError {
        throw fmt::SystemError(errno, "cannot open file '{}'", filename);
    \endrst
   */
-  SystemError(int error_code, CStringRef message) {
-    init(error_code, message, ArgList());
-  }
```
Comment (vitaut) [inline fmt/format.h]: Please keep this ctor.

File: fmt/format.h [hunk 7]
```diff
@@ -3446,9 +3537,6 @@ class WindowsError : public SystemError {
      }
    \endrst
   */
-  WindowsError(int error_code, CStringRef message) {
-    init(error_code, message, ArgList());
-  }
```
Comment (vitaut) [inline fmt/format.h]: And this

File: fmt/format.h (hunk 6 above)
Comment (starkmapper) [reply fmt/format.h]: It is generated by the FMT_VARIADIC_ macro.

File: fmt/format.h (hunk 7 above)
Comment (starkmapper) [reply fmt/format.h]: This one is also generated by the FMT_VARIADIC_ macro.

File: fmt/format.h (hunk 6 above)
Comment (vitaut) [reply fmt/format.h]: It's here for documentation purposes (note the apidoc above). If it is causing problems for bcc, I suggest conditionally compiling instead of removing.

File: fmt/format.h (hunk 5 above)
Comment (starkmapper) [reply fmt/format.h]: The only thing left of this block is the `FMT_MAKE_REF` macro. I moved those to have a single `#if FMT_USE_VARIADIC_TEMPLATES` block.

File: fmt/format.h (hunk 6 above)
Comment (starkmapper) [reply fmt/format.h]: It's not a bcc issue, it's because all `FMT_VARIADIC_*` macros now use the `FMT_VARIADIC_` macro, this normalizing variadic template handing, and less macro magic to maintain. The `FMT_VARIADIC_` macro generates functions with 0..15 templated arguments to mimic variadic templates, where the 0 argument one has the exact same signature as this manually defined one.
I propose to make it consistent with functions `printf`, `format` etc and add an overload with an extra `ArgList` parameter:
```cpp
SystemError(int error_code, CStringRef message, ArgList args) {
    init(error_code, message, args);
  }
```
This would be the best of both worlds: singe `FMT_VARIADIC_` macro and documentation (and more consistency as a bonus :-) ).

**Comment by starkmapper** [comment]
I'll make the proposed changes. Would you like me to squash again, or just push updates? Also, should I check the documentation using doxygen, or apidoc like you mentioned?

**Comment by starkmapper** [comment]
Adding the `..` didn't fix the appveyor build unfortunately.

**Comment by starkmapper** [comment]
> 
> 
> Adding the `..` didn't fix the appveyor build unfortunately. (https://ci.appveyor.com/project/vitaut/fmt/builds/23233738)

But adding `.` did, seems that adding the source path to the commandline became required in v3.13.

File: fmt/format.h (hunk 7 above)
Comment (starkmapper) [reply fmt/format.h]: This should be resolved with the newly added overload.

File: fmt/format.h (hunk 6 above)
Comment (vitaut) [reply fmt/format.h]: If this is not bcc issue, could you remove the macro changes from the current PR?

**Comment by vitaut** [comment]
> But adding . did, seems that adding the source path to the commandline became required in v3.13.

Thanks for fixing this.

> Would you like me to squash again, or just push updates?

Yes, please squash and remove the macro changes as discussed above.

> Also, should I check the documentation using doxygen, or apidoc like you mentioned?

I don't think it's necessary.

File: fmt/format.h (hunk 6 above)
Comment (starkmapper) [reply fmt/format.h]: The need for the macro changes *is* because of a bcc issue. If you feel very strongly about not changing the constructor signature in the documentation I can change the `FMT_VARIADIC_CTOR` macro not to generate the conflicting overload.

### Background

The original macro did c-style array initialization:
```cpp
ArgArray<2>::Type array = { MakeValue<BasicFormatter<char>>(v0), 
                            MakeValue<BasicFormatter<char>>(v1)};
```
Where `ArgArray<2>::Type` is an array of `Value`.
Somehow bcc tried to assign the results of `MakeValue<BasicFormatter<char>>(vN)` to the first member of `Value` (`int int_value` in the anonymous union) causing a "Cannot convert `MakeValue<BasicFormatter<char>>(vN) to int" error.

When I started fixing this issue I found the `FMT_VARIADIC_` function and started using that. This had already solved this issue by generating the individual assignment statements for the array:
```cpp
ArgArray<2>::Type arr;
arr[0] = MakeValue<BasicFormatter<char>>(v0), arr[1] = MakeValue<BasicFormatter<char>>(v1);
```

note: I left out namespaces for clarity

**Comment by starkmapper** [comment]
I changed the macros so the `WindowsError` and `SystemError` constructors can remain unchanged. Please let me know if there is anything else.

File: fmt/format.h (hunk 6 above)
Comment (vitaut) [reply fmt/format.h]: I see. The macro logic is fairly complicated and version 4.x is in maintenance mode, so I'd like the changes in this PR to be minimal, just enough for bcc32 compatibility. Right now it's hard to see what actually changed because the macros have been moved around. Could you return them to their original places and make changes there?

Comment (starkmapper) [reply fmt/format.h]: I started doing that, but found that the solution would depend on the (newer and more versatile) macros defined later on in the file. That is why I moved the large macros block up. To verify that I only moved them I used kdiff3 with manual alignment. I'll post those diffs so you can get a better idea of the changes I made.

Comment (starkmapper) [reply fmt/format.h]: Here you can see what I changed  to get this working for bcc with the existing macros. Not only achieving bcc compatibility, but also removing the second set of macro's, resulting in less "magic" to maintain:
Macro-diff

And the diff for the `FMT_MAKE_REF` macro's:
Macro-diff - old macros

The alignment of the diff is shown below. I marked the manual alignment markers with red. 
Macro-diff alignment

I hope this makes it clear that I only made minor changes, apart from moving the macros.

**Comment by vitaut** [comment]
Merged everything except the macro changes in https://github.com/fmtlib/fmt/commit/d2744bc848f63b222ed9bab3d93a9a3472f63eca. The macros need a bit more work because they are in their present location for a reason: to make the `system_header` pragma apply to them and not to the other parts of the code. Also it's not entirely clear to me how `FMT_VARIADIC_VAR_ARG` fixes anything.

**Comment by starkmapper** [comment]
I'll try to find a less invasive way to make the macros work.
